[Doc] Add custom dataset tutorial (#522)
open-mmlab/mmselfsup: docs/en/user_guides/4_pretrain_custom_dataset.md

# Tutorial 4: Pretrain with Custom Dataset

- [Tutorial 4: Pretrain with Custom Dataset](#tutorial-4-pretrain-with-custom-dataset)
  - [Train MAE on Custom Dataset](#train-mae-on-custom-dataset)
    - [Get the path of custom dataset](#get-the-path-of-custom-dataset)
    - [Choose one config as template](#choose-one-config-as-template)
    - [Edit the dataset related config](#edit-the-dataset-related-config)
  - [Train MAE on COCO Dataset](#train-mae-on-coco-dataset)
  - [Train SimCLR on Custom Dataset](#train-simclr-on-custom-dataset)
  - [Load Pre-trained Model to Speedup Convergence](#load-pre-trained-model-to-speedup-convergence)

In this tutorial, we provide some tips on how to conduct self-supervised learning on your own dataset(without the need of label).

## Train MAE on Custom Dataset

> In MMSelfSup, We support the `CustomDataset` from MMClassification,  which is able to read the images within the specified folder directly. You only need to prepare the path information of the custom dataset and edit the config.

Reviewer: Maybe we can add something like: similar to the ``ImageFolder`` in ``torchvision``.

### Get the path of custom dataset

> It should be like `/dataset/my_custom_dataset`

Reviewer 2: Better to keep consistent with config (`'data/custom_dataset/'`)

### Choose one config as template

> Here, we would like to use `configs/selfsup/mae/mae_vit-base-p16_8xb512-coslr-400e_in1k.py` as the example. We first copy this config file and rename it as `mae_vit-base-p16_8xb512-coslr-400e_custom_dataset.py`.

Reviewer: According to naming rules for config files, `mae_vit-base-p16_8xb512-coslr-400e_custom_dataset.py` should be renamed to `mae_vit-base-p16_8xb512-coslr-400e-${custom_dataset}.py`.

            'pos_embed': dict(decay_mult=0.),
            'mask_token': dict(decay_mult=0.),
            'cls_token': dict(decay_mult=0.)
        }))

# learning rate scheduler
param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=1e-4,
        by_epoch=True,
        begin=0,
        end=40,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingLR',
        T_max=360,
        by_epoch=True,
        begin=40,
        end=400,
        convert_to_iter_based=True)
]

# runtime settings
# pre-train for 400 epochs
train_cfg = dict(max_epochs=400)
default_hooks = dict(
    logger=dict(type='LoggerHook', interval=100),
    # only keeps the latest 3 checkpoints
    checkpoint=dict(type='CheckpointHook', interval=1, max_keep_ckpts=3))

# randomness
randomness = dict(seed=0, diff_rank_seed=True)
resume = True
```

### Edit the dataset related config

The dataset related config is defined in `'../_base_/datasets/imagenet_mae.py'` in `_base_`. We then copy the content of dataset config file into our created file `mae_vit-base-p16_8xb512-coslr-400e_custom_dataset.py`.

- Then we remove the `'../_base_/datasets/imagenet_mae.py'` in `_base_`.

> - Set the `dataset_type = 'mmcls.CustomDataset'`, and the path of the custom dataset ` dataset_type = /dataset/my_custom_dataset`.

Reviewer 2: ```suggestion
- Set the `dataset_type = 'mmcls.CustomDataset'`, and the path of the custom dataset ` data_root = data/custom_dataset/`.
```

> - Set the `dataset_type = 'mmcls.CustomDataset'`, and the path of the custom dataset ` dataset_type = /dataset/my_custom_dataset`.

Reviewer: Better to remind the user to add the ``custom_dataset`` in ``mmcls``, instead of ``mmselfsup``.

- Remove the `ann_file` in `train_dataloader`, and edit the `data_prefix` if needed.

> And edited the config will be like this:

Reviewer: ```suggestion
And the edited config will be like this:
```

    type='OptimWrapper',
    optimizer=optimizer,
    paramwise_cfg=dict(
        custom_keys={
            'ln': dict(decay_mult=0.0),
            'bias': dict(decay_mult=0.0),
            'pos_embed': dict(decay_mult=0.),
            'mask_token': dict(decay_mult=0.),
            'cls_token': dict(decay_mult=0.)
        }))

# learning rate scheduler
param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=1e-4,
        by_epoch=True,
        begin=0,
        end=40,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingLR',
        T_max=360,
        by_epoch=True,
        begin=40,
        end=400,
        convert_to_iter_based=True)
]

# runtime settings
# pre-train for 400 epochs
train_cfg = dict(max_epochs=400)
default_hooks = dict(
    logger=dict(type='LoggerHook', interval=100),
    # only keeps the latest 3 checkpoints
    checkpoint=dict(type='CheckpointHook', interval=1, max_keep_ckpts=3))

# randomness
randomness = dict(seed=0, diff_rank_seed=True)
resume = True
```

By using the edited config file, you are able to train a self-supervised model with MAE algorithm on the custom dataset.

## Train MAE on COCO Dataset

> 

Reviewer 2: Better to remind users to install MMDet and check the version at the beginning

optimizer = dict(
    type='AdamW', lr=1.5e-4 * 4096 / 256, betas=(0.9, 0.95), weight_decay=0.05)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
    paramwise_cfg=dict(
        custom_keys={
            'ln': dict(decay_mult=0.0),
            'bias': dict(decay_mult=0.0),
            'pos_embed': dict(decay_mult=0.),
            'mask_token': dict(decay_mult=0.),
            'cls_token': dict(decay_mult=0.)
        }))

# learning rate scheduler
param_scheduler = [
    dict(
        type='LinearLR',
        start_factor=1e-4,
        by_epoch=True,
        begin=0,
        end=40,
        convert_to_iter_based=True),
    dict(
        type='CosineAnnealingLR',
        T_max=360,
        by_epoch=True,
        begin=40,
        end=400,
        convert_to_iter_based=True)
]

# runtime settings
# pre-train for 400 epochs
train_cfg = dict(max_epochs=400)
default_hooks = dict(
    logger=dict(type='LoggerHook', interval=100),
    # only keeps the latest 3 checkpoints
    checkpoint=dict(type='CheckpointHook', interval=1, max_keep_ckpts=3))

# randomness
randomness = dict(seed=0, diff_rank_seed=True)
resume = True
```

## Train SimCLR on Custom Dataset

> We provide an example of using SimCLR on custom dataset, the main idea is similar with the [Train MAE on Custom Dataset

Reviewer: ```suggestion
We provide an example of using SimCLR on custom dataset, the main idea is similar to the [Train MAE on Custom Dataset
```

    collate_fn=dict(type='default_collate'),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        # ann_file='meta/train.txt',
        data_prefix=dict(img_path='./'),
        pipeline=train_pipeline))
# <<<<<<<<<<<<<<<<<<<<<< End of Changed <<<<<<<<<<<<<<<<<<<<<<<<<<<

# optimizer
optimizer = dict(type='LARS', lr=0.3, momentum=0.9, weight_decay=1e-6)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
    paramwise_cfg=dict(
        custom_keys={
            'bn': dict(decay_mult=0, lars_exclude=True),
            'bias': dict(decay_mult=0, lars_exclude=True),
            # bn layer in ResNet block downsample module
            'downsample.1': dict(decay_mult=0, lars_exclude=True),
        }))

# runtime settings
default_hooks = dict(
    # only keeps the latest 3 checkpoints
    checkpoint=dict(type='CheckpointHook', interval=10, max_keep_ckpts=3))

```

## Load pre-trained model to speedup convergence

To speedup the convergence of the model on your own dataset. You may use the pre-trained model as the initialization for the model's weight. You just need to specify the url of the pre-trained model via command.

```bash

> bash tools/dist_train.sh ${CONFIG} ${GPUS} --cfg-options ${PRETRAIN}

Reviewer 2: --cfg-options model.pretrained=......
